soroban-env-host/src/native_contract/base_types.rs

```diff
@@ -2,12 +2,120 @@ use crate::host::metered_clone::MeteredClone;
 use crate::host::{Host, HostError};
 
 use core::cmp::Ordering;
-use soroban_env_common::xdr::{AccountId, ScAddress};
+use soroban_env_common::xdr::{AccountId, ScAddress, ScErrorCode, ScErrorType};
 use soroban_env_common::{
     AddressObject, BytesObject, Compare, ConversionError, Env, EnvBase, MapObject, RawVal,
-    TryFromVal, VecObject,
+    StringObject, TryFromVal, VecObject,
 };
 
+#[derive(Clone)]
+pub struct String {
+    host: Host,
+    object: StringObject,
+}
+
+impl Compare<String> for Host {
+    type Error = HostError;
+
+    fn compare(&self, a: &String, b: &String) -> Result<Ordering, Self::Error> {
+        self.compare(&a.object.to_raw(), &b.object.to_raw())
+    }
+}
+
+impl TryFromVal<Host, StringObject> for String {
+    type Error = HostError;
+
+    fn try_from_val(env: &Host, val: &StringObject) -> Result<Self, Self::Error> {
+        Ok(String {
+            host: env.clone(),
+            object: *val,
+        })
+    }
+}
+
+impl TryFromVal<Host, RawVal> for String {
+    type Error = HostError;
+
+    fn try_from_val(env: &Host, val: &RawVal) -> Result<Self, Self::Error> {
+        let val = *val;
+        let obj: StringObject = val.try_into()?;
+        String::try_from_val(env, &obj)
+    }
+}
+
+impl TryFromVal<Host, String> for RawVal {
+    type Error = HostError;
+
+    fn try_from_val(_env: &Host, val: &String) -> Result<RawVal, Self::Error> {
+        Ok(val.object.into())
+    }
+}
+
+impl From<String> for StringObject {
+    fn from(b: String) -> Self {
+        b.object
+    }
+}
+
+impl String {
+    pub fn copy_to_rust_string(&self, env: &Host) -> Result<std::string::String, HostError> {
+        let len: u32 = env
+            .string_len(self.object)
+            .map_err(|_| ConversionError)?
+            .into();
+        let len = len as usize;
+        let mut vec = std::vec![0; len];
+        env.string_copy_to_slice(self.object, RawVal::U32_ZERO, &mut vec)
+            .map_err(|_| ConversionError)?;
+        std::string::String::from_utf8(vec).map_err(|_| ConversionError.into())
+    }
+
+    pub fn to_array_4(&self, env: &Host) -> Result<[u8; 4], HostError> {
```

Comment:
This is clunky, can't you use template here?

Reply:
Instead of a template, I'd use a const generic. Replace `4` with `N` and add a generic `N` as `const N: usize` for the function.

Reply:
Good point. Updated.